documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc

```diff
@@ -0,0 +1,232 @@
+= Infinispan {infinispanversion} Performance Guide
+The Infinispan community
+:toc2:
+:icons: font
+:toclevels: 3
+:numbered:
+:stem:
+
+This guide will give you information and tweaks about tuning Infinispan performance (both server and library mode).
+
+== Capacity planning
+
+Infinispan stores data in memory. Depending on operating mode data is stored in serialized form (server mode) or directly
+directly in memory (library mode). Data size can be estimated using sophisticated tools like  http://openjdk.java.net/projects/code-tools/jol/[Java Object Layout]
+and total amount of required memory can be calculated using the following formulas:
+
+[stem]
+.Total Data Set in library mode
+++++
+Total Data Set = Number Of Entries * (Key Size + Value Size + 200 b (Me tadata))
+++++
+
+[stem]
+.Total Data Set in server mode
+++++
+Total Data Set = Number Of Entries * (Serialized Key Size + Serialized Value Size + 200 b (Me tadata))
+++++
+
+
+In case of Local or Replicated mode, all data needs to fit in memory, so calculating the amount of required memory is trivial.
+
+Calculating memory requirements for Distributed mode is slightly more complicated and requires using the following:
+
+[stem]
+.Required memory in Distributed mode
+++++
+Required Memo ry = (Total Data Set)/(Nodes)*(1+(Node Failures)/(Nodes - Node Failures))
+++++
+
+Where:
+
+* `Total Data Set` - Estimated size of all data
+* `Nodes` - The number of nodes in the cluster
+* `Node Failures` - Number of possible failures (also `number of owners - 1`)
+
+Calculated amount of memory should be used for setting `Xmx` and `Xms` parameters.
+
+JVM as well as Infinispan require additional memory for other tasks like searches, allocating network buffers etc.
+It is advised to allocate no more than 50% of memory with living data.
+
+When considering large heaps, make sure there's enough CPU to perform garbage collection efficiently.
+
+== Java Virtual Machine settings
+
+Java Virtual Machine tuning might be divided into sections like memory or GC.
+Below is a list of helpful configuration parameters and a guide how to adjust them.
+
+=== Memory settings
+
+Adjusting memory size is one of the most crucial step in Infinispan tuning. The most commonly used JVM flags are:
+
+* `-Xms` - Defines the minimum heap size allowed.
+* `-Xmx` - Defines the maximum heap size allowed.
+* `-Xmn` - Defines the minimum and maximum value for the young generation.
+* `-XX:NewRatio` - Define the ratio between young and old generations. Should not be used if -Xmn is enabled.
+
+Using `Xms` equal to `Xmx` will prevent JVM from dynamically sizing memory and might decrease GC pauses cased by resizing.
```

Comment:
s/cased/caused/, but TBH I doubt resizing causes any GC pauses... starting with a smaller heap does mean you have a smaller young generation and more old GCs, but on the other hand those old GCs might be faster because the old gen is also smaller.